src/d3d12/d3d12_renderer.cpp

```diff
@@ -246,17 +234,52 @@ namespace wr
 
 	void D3D12RenderSystem::Render_MeshNode(MeshNode* node)
 	{
-		for (auto& mesh : node->m_model->m_meshes)
-		{
-			auto n_mesh = static_cast<D3D12Mesh*>(mesh);
-			auto frame_idx = GetFrameIdx();
+	}
+
+	//Render batches
+	void D3D12RenderSystem::RenderMeshBatches(SceneGraph& scene_graph)
+	{
+		
+		auto &batches = scene_graph.GetBatches();
 
-			auto d3d12_cb_handle = static_cast<D3D12ConstantBufferHandle*>(node->m_transform_cb);
+		//Update if it isn't already
 
-			d3d12::BindVertexBuffer(m_direct_cmd_list, n_mesh->m_vertex_buffer);
-			d3d12::BindConstantBuffer(m_direct_cmd_list, d3d12_cb_handle->m_native, 1, frame_idx);
+		bool should_update = batches.size() == 0;
+
+		for(auto &elem : batches)
+			if (elem.second.num_instances == 0) {
+				should_update = true;
+				break;
+			}
+
+		if(should_update)
+			scene_graph.Optimize();
+
+		//Render batches
+		for (auto &elem : scene_graph.GetBatches()) {
+
+			Model *model = elem.first;
+			temp::MeshBatch &batch = elem.second;
+
+			//Update object data
+			d3d12::UpdateConstantBuffer(batch.batchBuffer->m_native, GetFrameIdx(), &batch.data, sizeof(batch.data));
```

Comment:
Don't update constant buffers here. Should preferably happen somewhere that makes more sense (a place/time where also other buffers get updated).

Reply:
Moved to optimize